packages/melody-hooks/src/component.js

```diff
@@ -0,0 +1,181 @@
+/**
+ * Copyright 2018 trivago N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { enqueueComponent, options } from 'melody-idom';
+import { shallowEquals, shallowEqualsScalar } from './util/shallowEquals';
+import { HOOK_TYPE_USE_EFFECT, HOOK_TYPE_USE_REF } from './constants';
+import { setCurrentComponent, unsetCurrentComponent } from './util/hooks';
+
+const defaultComponentFn = props => props;
+
+function Component(element, componentFn) {
+    // part of the public API
+    this.el = element;
+
+    // needed for this type of component
+    this.componentFn = componentFn ? componentFn : defaultComponentFn;
+    this.hooks = [];
+    this.hooksPointer = -1;
+    this.isMounted = false;
+    this.props = null;
+    this.state = {};
+    this.data = null;
+
+    this.setState = this.setState.bind(this);
+}
+
+Object.assign(Component.prototype, {
+    /**
+     * Set new properties for the Component.
+     * This might cause the component to request an update.
+     */
+    apply(props) {
+        if (shallowEquals(props, this.props)) {
+            return;
+        }
+        this.props = props || null;
+        this.enqueueComponent();
+    },
+
+    /**
+     * Executed after a component has been mounted or updated.
+     * After this method has been triggered, the component is considered stable and
+     * accessing the DOM should be safe.
+     * The children of this Component might not have been rendered.
+     */
+    notify() {
+        const hooks = this.hooks;
+        for (let i = 0, l = hooks.length; i < l; i++) {
+            const hook = hooks[i];
+            const type = hook[0];
+
+            switch (type) {
+                case HOOK_TYPE_USE_EFFECT: {
+                    const dirty = hook[3];
+                    if (!dirty) continue;
+
+                    const callback = hook[1];
+                    const unsubscribe = hook[4];
+
+                    if (unsubscribe) unsubscribe();
+                    const unsubscribeNext = callback() || null;
```

Comment:
I think we should be a bit closer to React here.
This spot, executing within `notify` is actually `useLayoutEffect` since it fires after all changes have been made. That spot is going to cause recalculations but that's fine since they'd happen anyways.
`useMutationEffect` would need to be triggered from within `render` itself since that is the only spot where we can guarantee execution before siblings might be rendered.

And finally `effect` might need to be implemented by using `requestIdleCallback` or the same `window.postMessage` work-around we use in the render queue (https://github.com/trivago/melody/blob/master/packages/melody-idom/src/renderQueue.ts#L362) to execute after all DOM changes have been committed and layout has been recalculated as is stated in the React docs.

You'd also need to keep track of whether or not it has actually been executed so that you can flush if needed during `apply` or at the beginning of `render`.